### ISSUE TITLE
Missing possibly valid success_codes ?

### ISSUE THREAD

**Issue by webframp** [chef-cookbooks/powershell#40]
Running more than once with powershell4 recipe in the run_list gave:
```
Chef Client failed. 0 resources updated in 31.585417 seconds
[2015-02-03T12:38:38-08:00] FATAL: Mixlib::ShellOut::ShellCommandFailed: windows_package[Windows Management Framework Core4.0] (powershell::powershell4 line 36)
had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0, 42, 127, 3010], but received '2359302'
```

looking into it, 2359302 converts to 0x00240006 and is listed here:

http://support.microsoft.com/kb/938205

as a valid case meaning "0x00240006 WU_S_ALREADY_INSTALLED The update to be installed is already installed on the system". Perhaps this should also be included in the success_codes array?

**Comment by NimishaS** [comment]
@webframp , we have a guard condition in powershell4 recipe which checks if powershell4 is already installed on the machine or not:
```ruby
not_if do
    begin
        registry_data_exists?('HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine', { :name => 'PowerShellVersion', :type => :string, :data => '4.0' })
    rescue Chef::Exceptions::Win32RegKeyMissing
        false
    end
end
```
So ideally we should not run into this error code. Are you using master branch code or the released powershell cookbook?

**Comment by mattstratton** [comment]
I am using the master branch code and getting the error as well on Win2k8R2. Hmm.

**Comment by mattstratton** [comment]
Just checking into the registry - that guard isn't necessarily safe, as on my win2k8r2 machine the registry path for PowerShellEngine is `HKLM\SOFTWARE\Microsoft\PowerShell\1\PowerShellEngine`

There is no "3" in the path, just the "1".

Is version 3 a dependency of version 4?

**Comment by NimishaS** [comment]
@mattstratton , we'll work on this issue.

Version 3 is not a dependency of version 4. `HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine` is a registry path and `3` here has nothing to do with the powershell version. Version is determined by `:data => '4.0'` in the command below:
```ruby
registry_data_exists?('HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine', { :name => 'PowerShellVersion', :type => :string, :data => '4.0' })
```

**Comment by mattstratton** [comment]
It's odd - not every system has the /3 in the path, some have /1. Maybe we can wildcard match against that path. 

Sent from Outlook




On Wed, Feb 4, 2015 at 10:26 PM -0800, "NimishaS" <notifications@github.com> wrote:












@mattstratton , we'll work on this issue.



Version 3 is not a dependency of version 4. HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine is a registry path and 3 here has nothing to do with the powershell version. Version is determined by :data => '4.0' in the command below:

registry_data_exists?('HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine', { :name => 'PowerShellVersion', :type => :string, :data => '4.0' })

**Comment by NimishaS** [comment]
@mattstratton , we tried running powershell recipes on win2k8r2. For PS2, it had /1 registry path. But for PS3, PS4 and PS5, /3 was created. 
Did you try running powershell4 recipe when you observed that the registry path was /1?

**Comment by mattstratton** [comment]
We did. Version 4 recipe was the one we were running. 
Unfortunately I am not at that customer site anymore so I'm not sure how to reproduce it :)

Sent from Outlook




On Tue, Feb 10, 2015 at 4:22 AM -0800, "NimishaS" <notifications@github.com> wrote:












@mattstratton , we tried running powershell recipes on win2k8r2. For PS2, it had /1 registry path. But for PS3, PS4 and PS5, /3 was created. 

Did you try running powershell4 recipe when you observed that the registry path was /1?

**Comment by miguelaferreira** [comment]
I'm seeing a similar problem after having installed powershell 5.

The first time chef-client runs, it installs powershell version '5.0.10018.0'.
```
> $PSVersionTable.PSVersion

Major  Minor  Build  Revision
-----  -----  -----  --------
5      0      10018  0
```
Then, on the second time, it gets an error because the guard in the windows_package provider expects an older build version (5.0.9701.0)
```ruby
...
      not_if { registry_data_exists?('HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine', { :name => 'PowerShellVersion', :type => :string, :data => '5.0.9701.0' }) }
...
```

Full error:
 ```
Recipe: powershell::powershell5
  * windows_package[Windows Management Framework Core 5.0] action install[2015-06-09T06:46:27-07:00] INFO: Installing wi
ndows_package[Windows Management Framework Core 5.0] version latest

  Recipe: <Dynamically Defined Resource>
    * remote_file[c:/var/chef/cache/WindowsBlue-KB3037315-x64.msu] action create (up to date)
[2015-06-09T06:46:28-07:00] INFO: Starting installation...this could take awhile.

    ================================================================================
    Error executing action `install` on resource 'windows_package[Windows Management Framework Core 5.0]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    Expected process to exit with [0, 42, 127, 3010], but received '2359302'
    ---- Begin output of start "" /wait "c:\var\chef\cache\WindowsBlue-KB3037315-x64.msu" /quiet /norestart & exit %%ERR
ORLEVEL%% ----
    STDOUT:
    STDERR:
    ---- End output of start "" /wait "c:\var\chef\cache\WindowsBlue-KB3037315-x64.msu" /quiet /norestart & exit %%ERROR
LEVEL%% ----
    Ran start "" /wait "c:\var\chef\cache\WindowsBlue-KB3037315-x64.msu" /quiet /norestart & exit %%ERRORLEVEL%% returne
d 2359302

    Cookbook Trace:
    ---------------
    c:/var/chef/cache/cookbooks/windows/libraries/windows_package.rb:109:in `install_package'
    c:/var/chef/cache/cookbooks/windows/libraries/windows_package.rb:31:in `block in <class:WindowsCookbookPackage>'

    Resource Declaration:
    ---------------------
    # In c:/var/chef/cache/cookbooks/powershell/recipes/powershell5.rb

     36:     windows_package 'Windows Management Framework Core 5.0' do
     37:       source node['powershell']['powershell5']['url']
     38:       checksum node['powershell']['powershell5']['checksum']
     39:       installer_type :custom
     40:       options '/quiet /norestart'
     41:       action :install
     42:       success_codes [0, 42, 127, 3010]
     43:       # Note that the :immediately is to immediately notify the other resource,
     44:       # not to immediately reboot. The windows_reboot 'notifies' does that.
     45:       notifies :request, 'windows_reboot[powershell]', :immediately if reboot_pending? && node['powershell']['i
nstallation_reboot_mode'] != 'no_reboot'
     46:       not_if { registry_data_exists?('HKLM\SOFTWARE\Microsoft\PowerShell\3\PowerShellEngine', { :name => 'Power
ShellVersion', :type => :string, :data => '5.0.9701.0' }) }
     47:     end
     48:
     49:   else
     50:     Chef::Log.warn("PowerShell 5.0 is not supported or already installed on this version of Windows: #{node['pl
atform_version']}")
     51:   end
     52:

    Compiled Resource:
    ------------------
    # Declared in c:/var/chef/cache/cookbooks/powershell/recipes/powershell5.rb:36:in `from_file'

    windows_cookbook_package("Windows Management Framework Core 5.0") do
      provider Chef::Provider::WindowsCookbookPackage
      action [:install]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :windows_package
      cookbook_name "powershell"
      recipe_name "powershell5"
      source "http://download.microsoft.com/download/B/5/1/B5130F9A-6F07-481A-B4A6-CEDED7C96AE2/WindowsBlue-KB3037315-x6
4.msu"
      checksum "9a24de7b85fae96a87188a7fdaab27da3bc90d89426594cdc5f962234a61b7bd"
      installer_type :custom
      options "/quiet /norestart"
      success_codes [0, 42, 127, 3010]
      package_name "Windows Management Framework Core 5.0"
      timeout 600
      not_if { #code block }
    end
```

**Comment by dfduarte** [comment]
Hello Guys, good evening!

I have this same problem today, when I have used the Powershell 4 recipe installer (and 3, as well).

Following this guide: http://blogs.technet.com/b/askcore/archive/2011/02/15/how-to-use-dism-to-install-a-hotfix-from-within-windows.aspx, I extracted the MSU in order to debug any problems that can being ocour with the package.

When extracted, the MSU drops a file called PkgInstallOrder.txt, with some hotfixes which need to be installed with the PowerShell 4.  The MSU has the following hotfixes:

KB2809215; KB2872035; KB2872047; KB2819745. 

Following the DISM.exe instructions, the first two packages, have installed with success. The third package, returned a "ERROR 50", and this could be the root-cause of that MSU problem.

(BTW: the 4th package is the Powershell 4 itself, and it installed with success).

Maybe, this can be solved with some string to "force" of every already installed KB. I don't know if this is even possible, but, it's a way, I guess...

**Comment by dfduarte** [comment]
Possible fix:

Hello Again, guys. 

I figured out how to solve this issue (at least, it's worked for me): Verifying again the MSU package, I saw that it returned the error 0x80070422, and warning about a disabled service. Doing some research, I found this video: https://www.youtube.com/watch?v=YPAwpAXXKw8 that says which this problem is because the Windows Update service is disabled on the moment of installation.

I started the service, putting it at automatic, and I started again the chef-client. Now, the Powershell has installed successfully, without any errors.

Maybe so, it's just a question to verify if the "Windows Update" service is enabled at the time of installation, using the Windows_service resource from Chef: https://docs.chef.io/resource_windows_service.html
